state/migration_import.go

```diff
@@ -1407,35 +1407,45 @@ func (i *importer) makeApplicationDoc(a description.Application) (*applicationDo
 	return appDoc, nil
 }
 
+// makeCharmOrigin returns the charm origin for an application
+//
+// Previous versions of the Juju server and clients have treated applications charm
+// origins very loosely, particularly during switches. The server performed no validation
+// on origins recieved from the client, and client often mutated them incorrectly. For
+// instance, when switching from a ch charm to local, pylibjuju simply send back a copy
+// of the ch charm origin, whereas the CLI only set the source to local. Both resulted
+// in incorrect/invalidate origins.
+//
+// Calculate the origin Source and Revision from the charm url. Ensure ID, Hash and Channel
+// are dropped from local charm. If applicable, ID, Hash, Channel and Platform we have to
+// get from the origin. We can trust these since only a hacked client can break these.
```

Comment:
Or someone hand editing the db.

Reply:
Changed to "supported clients cannot break"